src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -138,11 +139,11 @@ module IncrementalBuildSyntaxTree =
                 )
             ), sourceRange, fileName, [||]
 
-        let parse _ =
-            let diagnosticsLogger = CompilationDiagnosticLogger("Parse", tcConfig.diagnosticsOptions)
-            // Return the disposable object that cleans up
-            use _holder = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.Parse)
-            try           
+        let getParseTask (source: FSharpSource) =
+            Threading.Tasks.Task.Run( fun () ->
```

Comment:
One thing which bothers me slighlty, is that now, we `lock` in some places inside incremental builder, which may lead to the thread pool exhaustion, and those tasks may not be executing.

This should help it a bit: https://github.com/dotnet/fsharp/pull/14904

Reply:
The lock is around a particular cache key, so it should be very granular, but I'll take a look.